filebeat/input/filestream/filestream.go

```diff
@@ -154,35 +151,18 @@ func (f *logFile) startFileMonitoringIfNeeded() {
 }
 
 func (f *logFile) closeIfTimeout(ctx unison.Canceler) {
-	timer := time.NewTimer(f.closeAfterInterval)
-	defer timer.Stop()
-
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case <-timer.C:
-			f.cancelReading()
-			return
-		}
+	if err := timed.Wait(ctx, f.closeAfterInterval); err == nil {
+		f.readerCtx.Cancel()
 	}
 }
 
 func (f *logFile) periodicStateCheck(ctx unison.Canceler) {
-	ticker := time.NewTicker(f.checkInterval)
-	defer ticker.Stop()
-
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case <-ticker.C:
-			if f.shouldBeClosed() {
-				f.cancelReading()
-				return
-			}
+	timed.Periodic(ctx, f.checkInterval, func() error {
+		if f.shouldBeClosed() {
+			f.readerCtx.Cancel()
 		}
-	}
+		return nil
+	})
```

Comment:
Replaced the for loop with `timed.Periodic`. The for loop did not check the state of the context before executing the select statement. Due to select checking channels in random order the `ctx.Done()` can be ignored multiple times in the worst case.